packages/yew-router/src/router.rs

```diff
@@ -72,16 +74,38 @@ fn base_router(props: &RouterProps) -> Html {
         basename,
     } = props.clone();
 
+    let basename = basename.map(|m| strip_slash_suffix(&m).to_string());
+    let navigator = Navigator::new(history.clone(), basename.clone());
+
+    let old_basename = use_mut_ref(|| Option::<String>::None);
+    let mut old_basename = old_basename.borrow_mut();
+    if basename != *old_basename {
+        // If `old_basename` is `Some`, path is probably prefixed with `old_basename`.
+        // If `old_basename` is `None`, path may or may not be prefixed with the new `basename`,
+        // depending on whether this is the first render.
+        let old_navigator = Navigator::new(
+            history.clone(),
+            old_basename.as_ref().or(basename.as_ref()).cloned(),
+        );
+        *old_basename = basename.clone();
+        let location = history.location();
+        let stripped = old_navigator.strip_basename(Cow::from(location.path()));
+        let prefixed = navigator.prefix_basename(&stripped);
+
+        if prefixed != location.path() {
+            history
+                .replace_with_query(prefixed, Raw(location.query_str()))
+                .unwrap_or_else(|never| match never {});
+        }
```

Comment:
I'm concerned about this flow
- basename changes
- history.replace_with_query is called here
- the router is rendered with old basename

there's nothing that stops the execution here to the router will be rendered as if no change happened.

Reply:
> the router is rendered with old basename

I don't recognize this step as something that happens. I assume that by "the router is rendered" you mean its children. Observe that the new `props.basename` is only visible to the router children via the `navigator` context. AFAICT the flow is like this:
1. router `basename` prop changes
2. BaseRouter is rendered
3. `navigator` is created with the (new) `props.basename`
4. `history.replace_with_query(path with old_basename -> path with props.basename)`
5. `navi_ctx` is created with `navigator`, still having the (new) `props.basename`

So the contents of the router should render with the correct (new) basename and only *after* the history has already been replaced.

Reply:
> I don't recognize this step as something that happens

The component execution will continue even after the `replace_with_query` is called. It behaves just a function! There's nothing changing the control flow, thus it'll finish the execution

Reply:
I edited my comment; my main confusion is not that the execution continues, but how exactly it continues with the wrong/old basename in the way you described. Currently investigating by logging the variables each render...

Reply:
Ok so what I did is insert these lines:
```
#[function_component(NavigationMenu)]
fn navigation_menu() -> Html {
    let navigator = use_navigator().unwrap();
    let location = use_location().unwrap();
    gloo::console::debug!(format!("rendering nav menu: base={:?} path={}", navigator.basename(), location.path()));
    ...
}
```
And here is the result of running just the `link_with_basename` test:
```
rendering nav menu: base=Some("/base") path=/base/
rendering nav menu: base=Some("/base") path=/base/
rendering nav menu: base=Some("/bayes") path=/bayes/
rendering nav menu: base=None path=/
rendering nav menu: base=Some("/bass") path=/bass/
```

Reply:
Thanks for the investigation. I was under the (wrong!) impression that the context holding old value after `replace_with_query` has been called would have an issue but it seems like it's not an issue. 

> If this resolves your concern, I will consider testing this equality so it doesn't regress.

Please do. A test that specifically triggers this conditional would be very helpful to have

```rust
        if prefixed != location.path() {
            history
                .replace_with_query(prefixed, Raw(location.query_str()))
                .unwrap_or_else(|never| match never {});
        }
```

Reply:
Cool, it now tests both the equality of basename and path prefix *and* a case in which that `if` isn't entered.